[BUG] [typescript-fetch] Type error in oneOf when legacyDiscriminatorBehavior is false

#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

<!--
Please follow the issue template below for bug reports.
Also please indicate in the issue title which language/library is concerned. Eg:  [BUG][JAVA] Bug generating foo with bar 
-->

##### Description

<!-- describe what is the question, suggestion or issue and why this is a problem for you. -->

##### openapi-generator version

<!-- which version of openapi-generator are you using, is it a regression? -->

##### OpenAPI declaration file content or url

```yaml
inputSpec: input.yaml
outputDir: test_client_ts
legacyDiscriminatorBehavior: false
```

```yaml
openapi: 3.0.1
info:
  title: ""
  version: ""
paths: {}
components:
  schemas:
    OneOf:
      discriminator:
        propertyName: type
        mapping":
          A: "#/components/schemas/A"
          B: "#/components/schemas/B"

    A:
      type: object
      properties:
        type:
          type: string

    B:
      type: object
      properties:
        type:
          type: string
```


```shell
> npm run build

test_client_ts/models/OneOf.ts:47:13 - error TS2322: Type 'A & { type: string; }' is not assignable to type 'OneOf'.
  Type 'A & { type: string; }' is not assignable to type '{ type: "B"; } & B'.
    Type 'A & { type: string; }' is not assignable to type '{ type: "B"; }'.
      Types of property 'type' are incompatible.
        Type 'string' is not assignable to type '"B"'.

47             return Object.assign({}, AFromJSONTyped(json, true), { type: 'A' });
               ~~~~~~

test_client_ts/models/OneOf.ts:49:13 - error TS2322: Type 'B & { type: string; }' is not assignable to type 'OneOf'.
  Type 'B & { type: string; }' is not assignable to type '{ type: "B"; } & B'.
    Type 'B & { type: string; }' is not assignable to type '{ type: "B"; }'.
      Types of property 'type' are incompatible.
        Type 'string' is not assignable to type '"B"'.

49             return Object.assign({}, BFromJSONTyped(json, true), { type: 'B' });
               ~~~~~~


Found 2 errors in the same file, starting at: test_client_ts/models/OneOf.ts:47
```

##### Generation Details

<!-- 
    Prefer CLI steps, including the language, libraries and various options. 
    Providing config-based settings allows for simpler testing across CLI and plugins. 
    For examples, see https://github.com/OpenAPITools/openapi-generator/tree/master/bin/configs
-->

##### Steps to reproduce

<!-- unambiguous set of steps to reproduce the bug.-->

##### Related issues/PRs

The change in #18154 now causes this issues.

##### Suggest a fix

--- a/modules/openapi-generator/src/main/resources/typescript-fetch/modelOneOf.mustache
+++ b/modules/openapi-generator/src/main/resources/typescript-fetch/modelOneOf.mustache
@@ -24,7 +24,7 @@ export function {{classname}}FromJSONTyped(json: any, ignoreDiscriminator: boole
     switch (json['{{discriminator.propertyBaseName}}']) {
 {{#discriminator.mappedModels}}
         case '{{mappingName}}':
-            return Object.assign({}, {{modelName}}FromJSONTyped(json, true), { {{discriminator.propertyName}}: '{{mappingName}}' });
+            return Object.assign({}, {{modelName}}FromJSONTyped(json, true), { {{discriminator.propertyName}}: '{{mappingName}}' } as const);
 {{/discriminator.mappedModels}}
         default:
             throw new Error(`No variant of {{classname}} exists with '{{discriminator.propertyName}}=${json['{{discriminator.propertyName}}']}'`);

